lib/Core/BuildEngine.cpp

```diff
@@ -1124,6 +1255,13 @@ class BuildEngineImpl {
       ruleInfo->setCancelled();
     }
 
+    for (auto& it: ruleInfos) {
+      // Cancel outstanding activity on rules
+      if (it.second.isScanning()) {
+        it.second.setCancelled();
+      }
```

Comment:
Can you explain why this is needed now? Was it not needed before, or it was an oversight?

Reply:
Also, this is really expensive because of the full scan... isn't there a cheaper way to simply iterate over the rules which are being scanned (`ruleInfosToScan`)?

Reply:
I believe it was an oversight.  It was exposed when I added a third pass to the cycle detection unit test, even before actually breaking the cycle.  i.e. trying to perform a build after cancellation caused asserts to fire.